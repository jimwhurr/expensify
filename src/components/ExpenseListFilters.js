import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState( () => ({ calendarFocused }));
    };

    onTextChange = (e) => {
            this.props.setTextFilter(e.target.value);
    };
    
    onSortChange =  (e) => {
        (e.target.value === 'amount') ? 
            this.props.sortByAmount() : 
                this.props.sortByDate()
    };

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            type="text"
                            className="text-input"
                            value={this.props.filters.text} 
                            onChange={ this.onTextChange }
                            placeholder="Search Filter"
                        />
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy} 
                            onChange={ this.onSortChange }>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={this.props.filters.startDate}
                            startDateId="startDate1"
                            endDate={this.props.filters.endDate}
                            endDateId="endDate1"
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            showClearDates={true}
                            displayFormat={'DD/MM/YYYY'}
                        />
                    </div>
                </div>
            </div>
        );
    }
};




const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);