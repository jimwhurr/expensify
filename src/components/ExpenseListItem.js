import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment'
import numeral from 'numeral';
import 'numeral/locales/en-gb';

numeral.locale('en-gb');

const ExpenseListItem = ( { id, description, amount, createdAt }) => (
    <div>
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3>
        <p>
        {numeral(amount/100).format('$0,0.00')}
        -
        {moment(createdAt).format('Do MMMM YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;