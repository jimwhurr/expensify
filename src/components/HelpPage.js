import React from 'react';

const HelpPage = () => (
    <div>
        Help me!
    </div>
);

export default HelpPage;