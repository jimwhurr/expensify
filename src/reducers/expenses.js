// -------------------------------------------------------------------------
//                     E X P E N S E S   R E D U C E R
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch(action.type) {

        case 'ADD_EXPENSE':     // REMEMBER we DON'T want to change the prev state!
            return [
                ...state,
                action.expense
            ];

        case 'REMOVE_EXPENSE':
            return state.filter( ({ id }) => id !== action.id );

        case 'EDIT_EXPENSE':
            return state.map( (expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates   // spread overrides
                    }
                }
                else {
                    return expense;
                }
            });

        case 'SET_EXPENSES':
            return action.expenses;
            
        default:
            return state;
    }
};
