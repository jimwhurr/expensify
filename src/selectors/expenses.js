import moment from 'moment';

// -------------------------------------------------------------------------
//                G E T   V I S I B L E   E X P E N S E S
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

export default (expenses, { text, sortBy, startDate, endDate }) => {
    // console.log('params: {', text, sortBy, startDate, endDate, '}');
    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ?  startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;
        
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;

    }).sort( (a, b) => {
        if (sortBy === 'date') {
            return (a.createdAt < b.createdAt) ? 1 : -1;
        }
        else if (sortBy === 'amount') {
            // console.log('a: ', a, ' b: ', b, a < b);
            return (a.amount < b.amount) ? 1 : -1;
        }
    });
};