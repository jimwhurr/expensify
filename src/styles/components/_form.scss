.form {
    display: flex;
    flex-direction: column;
    // VVV for all childre selector is >
    > * {
        margin-bottom: $m-size;
    }
}

.form__error {
    margin: 0 0 $m-size 0;
    font-style: italic;
}