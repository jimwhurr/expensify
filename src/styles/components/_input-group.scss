.input-group {
    // smaller screens
    display: flex;
    flex-direction: column;
    margin-bottom: $m-size;

    @media (min-width: $desktop-breakpoint) {
        // larger screens
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.input-group__item {
    margin-bottom: $s-size;

    @media (min-width: $desktop-breakpoint) {
        // larger screens
        margin: 0 $s-size 0 0;
    }
}