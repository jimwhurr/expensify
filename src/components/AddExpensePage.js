import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// switch over to class based component to avoid in-line ffuncttion
export class AddExpensePage extends React.Component {

    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    };

};

const mapDispatchToProps = (dispatch) => ({
    // set up props used to call dispatch
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);