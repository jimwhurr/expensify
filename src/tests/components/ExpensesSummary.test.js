import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

describe('ExpenseSummary', () => {

    test('should render expense summary for a single expense', () => {
        const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={195} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should render expense summary for multiple expenses', () => {
        const wrapper = shallow(<ExpensesSummary expenseCount={3} expensesTotal={156812} />);
        expect(wrapper).toMatchSnapshot();
    });
});