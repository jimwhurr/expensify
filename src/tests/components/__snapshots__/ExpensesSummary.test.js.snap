// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseSummary should render expense summary for a single expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    £1.95
  </p>
</div>
`;

exports[`ExpenseSummary should render expense summary for multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     
    expenses
     totalling 
    £1,568.12
  </p>
</div>
`;
